src/apps/socket/dev.lua

```diff
@@ -1,38 +1,58 @@
-module(...,package.seeall)
+module(..., package.seeall)
 
+local S = require("syscall")
+local h = require("syscall.helpers")
+local packet = require("core.packet")
+local bit = require("bit")
 local ffi = require("ffi")
 local C = ffi.C
 
-local packet = require("core.packet")
-require("lib.raw.raw_h")
-require("apps.socket.io_h")
-
-dev = {}
-
-function dev:new (ifname)
-   assert(ifname)
-   self.__index = self
-   local dev = {fd = C.open_raw(ifname)}
-   return setmetatable(dev, self)
+local c, t = S.c, S.types.t
+
+RawSocketDev = {}
+
+function RawSocketDev:new (name)
+   local tp = h.htons(c.ETH_P["ALL"])
+   local sock, err = S.socket(c.AF.PACKET, bit.bor(c.SOCK.RAW, c.SOCK.NONBLOCK), tp)
+   if not sock then return nil, err end
+   local index, err = S.util.if_nametoindex(name, sock)
+   if err then
+      sock:close()
+      return nil, err
+   end
+   local addr = t.sockaddr_ll{sll_family = c.AF.PACKET, sll_ifindex = index, sll_protocol = tp}
+   local ok, err = S.bind(sock, addr)
+   if not ok then
+      S.close(sock)
+      return nil, err
+   end
+   return setmetatable({sock = sock}, {__index = RawSocketDev})
 end
 
-function dev:transmit (p)
-   assert(self.fd)
-   assert(C.send_packet(self.fd, p) ~= -1)
+function RawSocketDev:transmit (p)
+   local _, err = S.write(self.sock, p.data, p.length)
```

Comment:
What happens if `write` fails to transmit `p.length` data? In `RawSocket:push` we barely call `self.dev:transmit(p)` and then unconditionally `packet.free(p)`.  I would suggest to retry sending until whole p.length is sent. Of course then we get branches in the Lua code, which could be better implemented in C for performance reasons (at least that is what Alex Gall did here https://github.com/SnabbCo/snabbswitch/pull/638)

As for the error handling, I can imagine a testcase where we send packets from a 82599 10Gbps port to a 1Gbps `eth0`. Then you can easily  get `EAGAIN` or `EWOULDBLOCK` (see `man 2 write`) for our non-blocking socket. I guess we could retry a couple of times before we "drop" the packet.